scipy/_lib/_uarray/_uarray_dispatch.cxx

```diff
@@ -35,7 +35,12 @@ class py_ref {
     return py_ref(object);
   }
 
-  ~py_ref() { Py_XDECREF(obj_); }
+  ~py_ref() {
+    // If python is already dead, leak the wrapped python objects
+    if (Py_IsInitialized()) {
+      Py_XDECREF(obj_);
+    }
```

Comment:
I think it would probably be better to go with Alban's other proposed strategy: don't use `py_ref` for global data with process lifetimes.

The `Py_IsInitialized()` check here will add some overhead to every usage of `py_ref` -- I'm not sure if that matters for scipy or not. It also seems a bit strange to me semantically: we need a valid thread state to call `Py_XDECREF` and the `Py_IsInitialized()` check guards some, but not all, of the cases where we don't have a valid thread state.

Reply:
@albanD would you agree to revert to https://github.com/scipy/scipy/commit/8cf845d7684bf5b051e060f622f6ace5d895ed17 (minus [90b5bf9](https://github.com/scipy/scipy/pull/21218/commits/90b5bf97eda4570a5156210fbf599830a5232821)) ?

Reply:
Yes that sound fair to me!

Reply:
[03c9799](https://github.com/scipy/scipy/pull/21218/commits/03c9799bbdd99ef906799de970aea0a49a285fcd) contains the final changes